In [1]:
import pandas as pd
import numpy as np

## Table of Content

#### 1. [The Basic](#content1)
#### 2. [Creating DataFrame](#content2)
#### 3. [Treating Null Values](#content3)
#### 4. [Modify/Add new colum(s)](#content4)
#### 5. [Deleting Columns](#content5)
#### 6. [Renaming Columns](#content6)
#### 7. [i Slicing DataFrame](#content7)
#### 7. [ii Slicing Using loc and iloc](#content77)
#### 8. [Adding a row](#content8)
#### 9. [Dropping rows](#content9)
#### 10. [Sorting](#content10)
#### 11. [Joins](#content11)
#### 12. [Groupby](#content12)

In [2]:
titanic = pd.read_csv('C:\data analytics\pandas_docs/train.csv')
df = titanic.copy()

<a id='content1'></a>
## 1. The Basic

In [9]:
df.head() # first 5 rows

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

In [10]:
df.tail() # last 5 rows

PassengerId  Survived  Pclass                                      Name  \
886          887         0       2                     Montvila, Rev. Juozas   
887          888         1       1              Graham, Miss. Margaret Edith   
888          889         0       3  Johnston, Miss. Catherine Helen "Carrie"   
889          890         1       1                     Behr, Mr. Karl Howell   
890          891         0       3                       Dooley, Mr. Patrick   

        Sex   Age  SibSp  Parch      Ticket   Fare Cabin Embarked  
886    male  27.0      0      0      211536  13.00   NaN        S  
887  female  19.0      0      0      112053  30.00   B42        S  
888  female   NaN      1      2  W./C. 6607  23.45   NaN        S  
889    male  26.0      0      0      111369  30.00  C148        C  
890    male  32.0      0      0      370376   7.75   NaN        Q

In [11]:
df.shape  # no.of rows and columns

(891, 12)

In [12]:
df.columns    # column names

Index(['PassengerId', 'Survived', 'Pclass', 'Name', 'Sex', 'Age', 'SibSp',
       'Parch', 'Ticket', 'Fare', 'Cabin', 'Embarked'],
      dtype='object')

In [15]:
df.index

RangeIndex(start=0, stop=891, step=1)

- Index no's starts from 0 and ends in 891
- Step 1 by 1  `eg(0,1,2,3.....,891)`

In [16]:
df['Pclass'].value_counts()  # Count of values in a particular column

Pclass
3    491
1    216
2    184
Name: count, dtype: int64

In [20]:
# General description of dataset
df.describe()

PassengerId    Survived      Pclass         Age       SibSp  \
count   891.000000  891.000000  891.000000  714.000000  891.000000   
mean    446.000000    0.383838    2.308642   29.699118    0.523008   
std     257.353842    0.486592    0.836071   14.526497    1.102743   
min       1.000000    0.000000    1.000000    0.420000    0.000000   
25%     223.500000    0.000000    2.000000   20.125000    0.000000   
50%     446.000000    0.000000    3.000000   28.000000    0.000000   
75%     668.500000    1.000000    3.000000   38.000000    1.000000   
max     891.000000    1.000000    3.000000   80.000000    8.000000   

            Parch        Fare  
count  891.000000  891.000000  
mean     0.381594   32.204208  
std      0.806057   49.693429  
min      0.000000    0.000000  
25%      0.000000    7.910400  
50%      0.000000   14.454200  
75%      0.000000   31.000000  
max      6.000000  512.329200

----

<a id='content2'></a>
## 2. Creating DataFrame

In [23]:
df_empty = pd.DataFrame()  # empty dataframe
df_empty.head()

Empty DataFrame
Columns: []
Index: []

In [25]:
student_dict = {'Name':['A','B','C'],'Age':[18,20,22],'Roll':[7,8,9]}
df_student = pd.DataFrame(student_dict)
df_student

Name  Age  Roll
0    A   18     7
1    B   20     8
2    C   22     9

In [28]:
df_student.reset_index()  # retrive a new index column

index Name  Age  Roll
0      0    A   18     7
1      1    B   20     8
2      2    C   22     9

-----

<a id='content3'></a>
## 3. Treating Null Values

In [4]:
df.isnull().sum()

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64

In [5]:
df['Age'].isnull().sum()

177

#### impute null values

In [6]:
df['Age'].fillna(df['Age'].mean(),inplace=True)

In [7]:
df['Age'].isnull().sum()

0

In [8]:
df['Sex'].fillna(df['Sex'].mode(),inplace=True)

In [9]:
df['Sex'].isnull().sum()

0

-----

<a id='content4'></a>
## 4. Modify/Add new colum(s)

In [10]:
df.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

In [3]:
df['Sex'] = df['Sex'].map({'male':'0','female':'1'})

In [4]:
df.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name Sex   Age  SibSp  Parch  \
0                            Braund, Mr. Owen Harris   0  22.0      1      0   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...   1  38.0      1      0   
2                             Heikkinen, Miss. Laina   1  26.0      0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)   1  35.0      1      0   
4                           Allen, Mr. William Henry   0  35.0      0      0   

             Ticket     Fare Cabin Embarked  
0         A/5 21171   7.2500   NaN        S  
1          PC 17599  71.2833   C85        C  
2  STON/O2. 3101282   7.9250   NaN        S  
3            113803  53.1000  C123        S  
4            373450   8.0500   NaN        S

In [15]:
# finding last and first name from Name column
df['last_name'] = df['Name'].apply(lambda x: x.split(',')[0])
df['first_name'] = df['Name'].apply(lambda x: x.split(',')[1])
#df['first_name']=df['Name'].apply(lambda x: ' '.join(x.split(',')[1:]))

In [16]:
df.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name Sex   Age  SibSp  Parch  \
0                            Braund, Mr. Owen Harris   0  22.0      1      0   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...   1  38.0      1      0   
2                             Heikkinen, Miss. Laina   1  26.0      0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)   1  35.0      1      0   
4                           Allen, Mr. William Henry   0  35.0      0      0   

             Ticket     Fare Cabin Embarked  last_name  \
0         A/5 21171   7.2500   NaN        S     Braund   
1          PC 17599  71.2833   C85        C    Cumings   
2  STON/O2. 3101282   7.9250   NaN        S  Heikkinen   
3            113803  53.1000  C123        S   Futrelle   
4            373450   8.0500   NaN        S      Allen   

                                    first_name  
0                              Mr. Owen Harris  
1   Mrs. John Bradley (Florence Briggs Thayer)  
2                                  Miss. Laina  
3           Mrs. Jacques Heath (Lily May Peel)  
4                            Mr. William Henry

In [21]:
# Sets to 1 for men in 3rd class
df['Third&Men'] = df.apply(lambda x: int(x['Pclass']==3 and x['Sex']=='0'),axis=1)

- If we remove the `int` o/p will be boolean

In [22]:
df.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name Sex   Age  SibSp  Parch  \
0                            Braund, Mr. Owen Harris   0  22.0      1      0   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...   1  38.0      1      0   
2                             Heikkinen, Miss. Laina   1  26.0      0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)   1  35.0      1      0   
4                           Allen, Mr. William Henry   0  35.0      0      0   

             Ticket     Fare Cabin Embarked  last_name  \
0         A/5 21171   7.2500   NaN        S     Braund   
1          PC 17599  71.2833   C85        C    Cumings   
2  STON/O2. 3101282   7.9250   NaN        S  Heikkinen   
3            113803  53.1000  C123        S   Futrelle   
4            373450   8.0500   NaN        S      Allen   

                                    first_name  Third&Men  
0                              Mr. Owen Harris          1  
1   Mrs. John Bradley (Florence Briggs Thayer)          0  
2                                  Miss. Laina          0  
3           Mrs. Jacques Heath (Lily May Peel)          0  
4                            Mr. William Henry          1

##### Calling some custom function

In [23]:
def findAgeGroup(age):
    if age<18:
        return 1
    elif age>=18 and age<40:
        return 2
    elif age>=40 and age<60:
        return 3
    else:
        return 4
df['Age_Group'] = df['Age'].apply(lambda x : findAgeGroup(x))
# Calling a custom function

In [24]:
df.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name Sex   Age  SibSp  Parch  \
0                            Braund, Mr. Owen Harris   0  22.0      1      0   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...   1  38.0      1      0   
2                             Heikkinen, Miss. Laina   1  26.0      0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)   1  35.0      1      0   
4                           Allen, Mr. William Henry   0  35.0      0      0   

             Ticket     Fare Cabin Embarked  last_name  \
0         A/5 21171   7.2500   NaN        S     Braund   
1          PC 17599  71.2833   C85        C    Cumings   
2  STON/O2. 3101282   7.9250   NaN        S  Heikkinen   
3            113803  53.1000  C123        S   Futrelle   
4            373450   8.0500   NaN        S      Allen   

                                    first_name  Third&Men  Age_Group  
0                              Mr. Owen Harris          1          2  
1   Mrs. John Bradley (Florence Briggs Thayer)          0          2  
2                                  Miss. Laina          0          2  
3           Mrs. Jacques Heath (Lily May Peel)          0          2  
4                            Mr. William Henry          1          2

----

<a id='content5'></a>
## 5. Deleting Columns

In [29]:
# dropping the column passenger id
df.drop(['PassengerId'],axis=1,inplace=True)

In [30]:
df.head()

Survived  Pclass                                               Name Sex  \
0         0       3                            Braund, Mr. Owen Harris   0   
1         1       1  Cumings, Mrs. John Bradley (Florence Briggs Th...   1   
2         1       3                             Heikkinen, Miss. Laina   1   
3         1       1       Futrelle, Mrs. Jacques Heath (Lily May Peel)   1   
4         0       3                           Allen, Mr. William Henry   0   

    Age  SibSp  Parch            Ticket     Fare Cabin Embarked  last_name  \
0  22.0      1      0         A/5 21171   7.2500   NaN        S     Braund   
1  38.0      1      0          PC 17599  71.2833   C85        C    Cumings   
2  26.0      0      0  STON/O2. 3101282   7.9250   NaN        S  Heikkinen   
3  35.0      1      0            113803  53.1000  C123        S   Futrelle   
4  35.0      0      0            373450   8.0500   NaN        S      Allen   

                                    first_name  Third&Men  Age_Group  
0                              Mr. Owen Harris          1          2  
1   Mrs. John Bradley (Florence Briggs Thayer)          0          2  
2                                  Miss. Laina          0          2  
3           Mrs. Jacques Heath (Lily May Peel)          0          2  
4                            Mr. William Henry          1          2

---

<a id='content6'></a>
## 6. Renaming Columns

In [32]:
# Renaming some columns
df.rename(columns={'Sex':'Gender','Name':'Full_Name','last_name':'Surname','first_name':'Name'},inplace=True)

In [33]:
df.head()

Survived  Pclass                                          Full_Name Gender  \
0         0       3                            Braund, Mr. Owen Harris      0   
1         1       1  Cumings, Mrs. John Bradley (Florence Briggs Th...      1   
2         1       3                             Heikkinen, Miss. Laina      1   
3         1       1       Futrelle, Mrs. Jacques Heath (Lily May Peel)      1   
4         0       3                           Allen, Mr. William Henry      0   

    Age  SibSp  Parch            Ticket     Fare Cabin Embarked    Surname  \
0  22.0      1      0         A/5 21171   7.2500   NaN        S     Braund   
1  38.0      1      0          PC 17599  71.2833   C85        C    Cumings   
2  26.0      0      0  STON/O2. 3101282   7.9250   NaN        S  Heikkinen   
3  35.0      1      0            113803  53.1000  C123        S   Futrelle   
4  35.0      0      0            373450   8.0500   NaN        S      Allen   

                                          Name  Third&Men  Age_Group  
0                              Mr. Owen Harris          1          2  
1   Mrs. John Bradley (Florence Briggs Thayer)          0          2  
2                                  Miss. Laina          0          2  
3           Mrs. Jacques Heath (Lily May Peel)          0          2  
4                            Mr. William Henry          1          2

<a id='content7'></a>
## 7.`i` Slicing DataFrame

In [34]:
third_class = df[df['Pclass']==3]

In [38]:
third_class.reset_index(drop=True).head()  # drop=True it actually adds a index column rather

Survived  Pclass                       Full_Name Gender   Age  SibSp  \
0         0       3         Braund, Mr. Owen Harris      0  22.0      1   
1         1       3          Heikkinen, Miss. Laina      1  26.0      0   
2         0       3        Allen, Mr. William Henry      0  35.0      0   
3         0       3                Moran, Mr. James      0   NaN      0   
4         0       3  Palsson, Master. Gosta Leonard      0   2.0      3   

   Parch            Ticket     Fare Cabin Embarked    Surname  \
0      0         A/5 21171   7.2500   NaN        S     Braund   
1      0  STON/O2. 3101282   7.9250   NaN        S  Heikkinen   
2      0            373450   8.0500   NaN        S      Allen   
3      0            330877   8.4583   NaN        Q      Moran   
4      1            349909  21.0750   NaN        S    Palsson   

                     Name  Third&Men  Age_Group  
0         Mr. Owen Harris          1          2  
1             Miss. Laina          0          2  
2       Mr. William Henry          1          2  
3               Mr. James          1          4  
4   Master. Gosta Leonard          1          1

- Females with age>60

In [39]:
df[(df['Age']>60) & (df['Gender']=='1')]

Survived  Pclass                                  Full_Name Gender   Age  \
275         1       1          Andrews, Miss. Kornelia Theodosia      1  63.0   
483         1       3                     Turkula, Mrs. (Hedwig)      1  63.0   
829         1       1  Stone, Mrs. George Nelson (Martha Evelyn)      1  62.0   

     SibSp  Parch  Ticket     Fare Cabin Embarked  Surname  \
275      1      0   13502  77.9583    D7        S  Andrews   
483      0      0    4134   9.5875   NaN        S  Turkula   
829      0      0  113572  80.0000   B28      NaN    Stone   

                                    Name  Third&Men  Age_Group  
275             Miss. Kornelia Theodosia          0          4  
483                        Mrs. (Hedwig)          0          4  
829   Mrs. George Nelson (Martha Evelyn)          0          4

In [40]:
df1=df[['Age','Pclass','Gender']]
df1.head()

Age  Pclass Gender
0  22.0       3      0
1  38.0       1      1
2  26.0       3      1
3  35.0       1      1
4  35.0       3      0

In [43]:
# Selecting numercial columns only
numerics = ['int16','int32','int64','float16','float32','float64']

df_num = df.select_dtypes(include=numerics)
df_num.head()

Survived  Pclass   Age  SibSp  Parch     Fare  Third&Men  Age_Group
0         0       3  22.0      1      0   7.2500          1          2
1         1       1  38.0      1      0  71.2833          0          2
2         1       3  26.0      0      0   7.9250          0          2
3         1       1  35.0      1      0  53.1000          0          2
4         0       3  35.0      0      0   8.0500          1          2

In [44]:
df_cat = df.select_dtypes(include=['object'])
df_cat.head()

Full_Name Gender            Ticket  \
0                            Braund, Mr. Owen Harris      0         A/5 21171   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...      1          PC 17599   
2                             Heikkinen, Miss. Laina      1  STON/O2. 3101282   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)      1            113803   
4                           Allen, Mr. William Henry      0            373450   

  Cabin Embarked    Surname                                         Name  
0   NaN        S     Braund                              Mr. Owen Harris  
1   C85        C    Cumings   Mrs. John Bradley (Florence Briggs Thayer)  
2   NaN        S  Heikkinen                                  Miss. Laina  
3  C123        S   Futrelle           Mrs. Jacques Heath (Lily May Peel)  
4   NaN        S      Allen                            Mr. William Henry

<a id='content77'></a>
## 7.`ii` Slicing Using loc and iloc

### iloc

In [45]:
# first 100 rows & all columns
df_sub1 = df.iloc[0:100,:]
df_sub1

Survived  Pclass                                          Full_Name  \
0          0       3                            Braund, Mr. Owen Harris   
1          1       1  Cumings, Mrs. John Bradley (Florence Briggs Th...   
2          1       3                             Heikkinen, Miss. Laina   
3          1       1       Futrelle, Mrs. Jacques Heath (Lily May Peel)   
4          0       3                           Allen, Mr. William Henry   
..       ...     ...                                                ...   
95         0       3                        Shorney, Mr. Charles Joseph   
96         0       1                          Goldschmidt, Mr. George B   
97         1       1                    Greenfield, Mr. William Bertram   
98         1       2               Doling, Mrs. John T (Ada Julia Bone)   
99         0       2                                  Kantor, Mr. Sinai   

   Gender   Age  SibSp  Parch            Ticket     Fare    Cabin Embarked  \
0       0  22.0      1      0         A/5 21171   7.2500      NaN        S   
1       1  38.0      1      0          PC 17599  71.2833      C85        C   
2       1  26.0      0      0  STON/O2. 3101282   7.9250      NaN        S   
3       1  35.0      1      0            113803  53.1000     C123        S   
4       0  35.0      0      0            373450   8.0500      NaN        S   
..    ...   ...    ...    ...               ...      ...      ...      ...   
95      0   NaN      0      0            374910   8.0500      NaN        S   
96      0  71.0      0      0          PC 17754  34.6542       A5        C   
97      0  23.0      0      1          PC 17759  63.3583  D10 D12        C   
98      1  34.0      0      1            231919  23.0000      NaN        S   
99      0  34.0      1      0            244367  26.0000      NaN        S   

        Surname                                         Name  Third&Men  \
0        Braund                              Mr. Owen Harris          1   
1       Cumings   Mrs. John Bradley (Florence Briggs Thayer)          0   
2     Heikkinen                                  Miss. Laina          0   
3      Futrelle           Mrs. Jacques Heath (Lily May Peel)          0   
4         Allen                            Mr. William Henry          1   
..          ...                                          ...        ...   
95      Shorney                           Mr. Charles Joseph          1   
96  Goldschmidt                                 Mr. George B          0   
97   Greenfield                          Mr. William Bertram          0   
98       Doling                 Mrs. John T (Ada Julia Bone)          0   
99       Kantor                                    Mr. Sinai          0   

    Age_Group  
0           2  
1           2  
2           2  
3           2  
4           2  
..        ...  
95          4  
96          4  
97          2  
98          2  
99          2  

[100 rows x 15 columns]

In [46]:
df_sub2 = df.iloc[:250,[1,8]]
df_sub2  #Returns first 250 rows and columns at those indices in df.columns

Pclass     Fare
0         3   7.2500
1         1  71.2833
2         3   7.9250
3         1  53.1000
4         3   8.0500
..      ...      ...
245       1  90.0000
246       3   7.7750
247       2  14.5000
248       1  52.5542
249       2  26.0000

[250 rows x 2 columns]

### loc

In [47]:
# first 500 rows
df_sub3 = df.loc[:500,:]

In [49]:
# Gender and age of age>50
df_sub4 = df.loc[(df['Age']>50),['Gender','Age']]
df_sub4

Gender   Age
6        0  54.0
11       1  58.0
15       1  55.0
33       0  66.0
54       0  65.0
..     ...   ...
820      1  52.0
829      1  62.0
851      0  74.0
857      0  51.0
879      1  56.0

[64 rows x 2 columns]

---

<a id='content8'></a>
## 8. Adding a row

In [51]:
# Adding row using 'append' function
row=dict({'Age':24,'Full_Name':'Sinan','Survived':'Y'})
df=df.append(row,ignore_index=True)

AttributeError: 'DataFrame' object has no attribute 'append'

In [52]:
df.loc[len(df.index)]=row
df.tail()

Survived  Pclass                                 Full_Name Gender   Age  \
887        1     1.0              Graham, Miss. Margaret Edith      1  19.0   
888        0     3.0  Johnston, Miss. Catherine Helen "Carrie"      1   NaN   
889        1     1.0                     Behr, Mr. Karl Howell      0  26.0   
890        0     3.0                       Dooley, Mr. Patrick      0  32.0   
891        Y     NaN                                     Sinan    NaN  24.0   

     SibSp  Parch      Ticket   Fare Cabin Embarked   Surname  \
887    0.0    0.0      112053  30.00   B42        S    Graham   
888    1.0    2.0  W./C. 6607  23.45   NaN        S  Johnston   
889    0.0    0.0      111369  30.00  C148        C      Behr   
890    0.0    0.0      370376   7.75   NaN        Q    Dooley   
891    NaN    NaN         NaN    NaN   NaN      NaN       NaN   

                                Name  Third&Men  Age_Group  
887             Miss. Margaret Edith        0.0        2.0  
888   Miss. Catherine Helen "Carrie"        0.0        4.0  
889                  Mr. Karl Howell        0.0        2.0  
890                      Mr. Patrick        1.0        2.0  
891                              NaN        NaN        NaN

---

<a id='content9'></a>
## 9. Dropping rows

In [53]:
df.drop(df.index[-1])

Survived  Pclass                                          Full_Name  \
0          0     3.0                            Braund, Mr. Owen Harris   
1          1     1.0  Cumings, Mrs. John Bradley (Florence Briggs Th...   
2          1     3.0                             Heikkinen, Miss. Laina   
3          1     1.0       Futrelle, Mrs. Jacques Heath (Lily May Peel)   
4          0     3.0                           Allen, Mr. William Henry   
..       ...     ...                                                ...   
886        0     2.0                              Montvila, Rev. Juozas   
887        1     1.0                       Graham, Miss. Margaret Edith   
888        0     3.0           Johnston, Miss. Catherine Helen "Carrie"   
889        1     1.0                              Behr, Mr. Karl Howell   
890        0     3.0                                Dooley, Mr. Patrick   

    Gender   Age  SibSp  Parch            Ticket     Fare Cabin Embarked  \
0        0  22.0    1.0    0.0         A/5 21171   7.2500   NaN        S   
1        1  38.0    1.0    0.0          PC 17599  71.2833   C85        C   
2        1  26.0    0.0    0.0  STON/O2. 3101282   7.9250   NaN        S   
3        1  35.0    1.0    0.0            113803  53.1000  C123        S   
4        0  35.0    0.0    0.0            373450   8.0500   NaN        S   
..     ...   ...    ...    ...               ...      ...   ...      ...   
886      0  27.0    0.0    0.0            211536  13.0000   NaN        S   
887      1  19.0    0.0    0.0            112053  30.0000   B42        S   
888      1   NaN    1.0    2.0        W./C. 6607  23.4500   NaN        S   
889      0  26.0    0.0    0.0            111369  30.0000  C148        C   
890      0  32.0    0.0    0.0            370376   7.7500   NaN        Q   

       Surname                                         Name  Third&Men  \
0       Braund                              Mr. Owen Harris        1.0   
1      Cumings   Mrs. John Bradley (Florence Briggs Thayer)        0.0   
2    Heikkinen                                  Miss. Laina        0.0   
3     Futrelle           Mrs. Jacques Heath (Lily May Peel)        0.0   
4        Allen                            Mr. William Henry        1.0   
..         ...                                          ...        ...   
886   Montvila                                  Rev. Juozas        0.0   
887     Graham                         Miss. Margaret Edith        0.0   
888   Johnston               Miss. Catherine Helen "Carrie"        0.0   
889       Behr                              Mr. Karl Howell        0.0   
890     Dooley                                  Mr. Patrick        1.0   

     Age_Group  
0          2.0  
1          2.0  
2          2.0  
3          2.0  
4          2.0  
..         ...  
886        2.0  
887        2.0  
888        4.0  
889        2.0  
890        2.0  

[891 rows x 15 columns]

---

<a id='content10'></a>
## 10. Sorting

In [56]:
# Sorting the Age column by decreasing order
#df['Age']=df['Age'].sort_values(ascending=False)
df = df.sort_values(by=['Age'],ascending=False)
df.head()

Survived  Pclass                             Full_Name Gender   Age  \
630        1     1.0  Barkworth, Mr. Algernon Henry Wilson      0  80.0   
851        0     3.0                   Svensson, Mr. Johan      0  74.0   
96         0     1.0             Goldschmidt, Mr. George B      0  71.0   
493        0     1.0               Artagaveytia, Mr. Ramon      0  71.0   
116        0     3.0                  Connors, Mr. Patrick      0  70.5   

     SibSp  Parch    Ticket     Fare Cabin Embarked       Surname  \
630    0.0    0.0     27042  30.0000   A23        S     Barkworth   
851    0.0    0.0    347060   7.7750   NaN        S      Svensson   
96     0.0    0.0  PC 17754  34.6542    A5        C   Goldschmidt   
493    0.0    0.0  PC 17609  49.5042   NaN        C  Artagaveytia   
116    0.0    0.0    370369   7.7500   NaN        Q       Connors   

                           Name  Third&Men  Age_Group  
630   Mr. Algernon Henry Wilson        0.0        4.0  
851                   Mr. Johan        1.0        4.0  
96                 Mr. George B        0.0        4.0  
493                   Mr. Ramon        0.0        4.0  
116                 Mr. Patrick        1.0        4.0

<a id='content11'></a>
## 11. Joins

Create two separate data frames and try all 5 (Cross,Inner, Left, Right,and Outer Joins)

In [3]:
sno = [i+1 for i in range(100)]
marks = np.random.randint(100,size=100)
print(len(marks))
marks_df = pd.DataFrame({'Sno':sno,'Marks':marks})
marks_df

100


Sno  Marks
0     1      4
1     2     29
2     3     46
3     4      8
4     5     63
..  ...    ...
95   96     19
96   97     33
97   98     64
98   99     60
99  100     43

[100 rows x 2 columns]

In [4]:
sno = [i+1 for i in range(100)]
age = np.random.randint(100,size=100)
print(len(age))
age_df = pd.DataFrame({'Sno':sno,'Age':age})
age_df

100


Sno  Age
0     1   15
1     2    5
2     3   67
3     4   37
4     5   97
..  ...  ...
95   96   22
96   97   42
97   98   37
98   99   24
99  100   41

[100 rows x 2 columns]

#### Cross Join

In [5]:
cross_join = pd.merge(marks_df,age_df,how='cross')
cross_join

Sno_x  Marks  Sno_y  Age
0         1      4      1   15
1         1      4      2    5
2         1      4      3   67
3         1      4      4   37
4         1      4      5   97
...     ...    ...    ...  ...
9995    100     43     96   22
9996    100     43     97   42
9997    100     43     98   37
9998    100     43     99   24
9999    100     43    100   41

[10000 rows x 4 columns]

#### Inner Join

In [6]:
inner_join = pd.merge(age_df,marks_df,how='inner')
print(inner_join.shape)
inner_join

(100, 3)


Sno  Age  Marks
0     1   15      4
1     2    5     29
2     3   67     46
3     4   37      8
4     5   97     63
..  ...  ...    ...
95   96   22     19
96   97   42     33
97   98   37     64
98   99   24     60
99  100   41     43

[100 rows x 3 columns]

#### Left and Right Outer Joins

Adding more rows

In [7]:
age_df.loc[len(age_df.index)]={'Sno':101,'Age':22}
age_df.loc[len(age_df.index)]={'Sno':102,'Age':28}
age_df.loc[len(age_df.index)]={'Sno':103,'Age':30}
age_df.loc[len(age_df.index)]={'Sno':104,'Age':33}
age_df.loc[len(age_df.index)]={'Sno':105,'Age':45}

#### Left Join

In [8]:
left_join = pd.merge(age_df,marks_df,how='left',on='Sno')
print(left_join.shape)
left_join.tail()

(105, 3)


Sno  Age  Marks
100  101   22    NaN
101  102   28    NaN
102  103   30    NaN
103  104   33    NaN
104  105   45    NaN

#### Right Join

In [11]:
right_join = pd.merge(marks_df,age_df,how='right',on='Sno')
print(right_join.shape)
right_join.tail()

(105, 3)


Sno  Marks  Age
100  101    NaN   22
101  102    NaN   28
102  103    NaN   30
103  104    NaN   33
104  105    NaN   45

#### Full Outer Join

Modify the marks frame

In [14]:
marks_df.loc[len(marks_df.index)]={'Sno':106,'Marks':77}
marks_df.tail()

Sno  Marks
96    97     33
97    98     64
98    99     60
99   100     43
100  106     77

#### outer join

In [16]:
out_join = pd.merge(marks_df,age_df,how='outer')
print(out_join.shape)
out_join.tail(10)

(106, 3)


Sno  Marks   Age
96    97   33.0  42.0
97    98   64.0  37.0
98    99   60.0  24.0
99   100   43.0  41.0
100  106   77.0   NaN
101  101    NaN  22.0
102  102    NaN  28.0
103  103    NaN  30.0
104  104    NaN  33.0
105  105    NaN  45.0

<a id='content12'></a>
## 12. Groupby

In [17]:
df.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

In [18]:
groups = df.groupby(['Pclass'])
groups.get_group(1)

PassengerId  Survived  Pclass  \
1              2         1       1   
3              4         1       1   
6              7         0       1   
11            12         1       1   
23            24         1       1   
..           ...       ...     ...   
871          872         1       1   
872          873         0       1   
879          880         1       1   
887          888         1       1   
889          890         1       1   

                                                  Name     Sex   Age  SibSp  \
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
6                              McCarthy, Mr. Timothy J    male  54.0      0   
11                            Bonnell, Miss. Elizabeth  female  58.0      0   
23                        Sloper, Mr. William Thompson    male  28.0      0   
..                                                 ...     ...   ...    ...   
871   Beckwith, Mrs. Richard Leonard (Sallie Monypeny)  female  47.0      1   
872                           Carlsson, Mr. Frans Olof    male  33.0      0   
879      Potter, Mrs. Thomas Jr (Lily Alexenia Wilson)  female  56.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
889                              Behr, Mr. Karl Howell    male  26.0      0   

     Parch    Ticket     Fare        Cabin Embarked  
1        0  PC 17599  71.2833          C85        C  
3        0    113803  53.1000         C123        S  
6        0     17463  51.8625          E46        S  
11       0    113783  26.5500         C103        S  
23       0    113788  35.5000           A6        S  
..     ...       ...      ...          ...      ...  
871      1     11751  52.5542          D35        S  
872      0       695   5.0000  B51 B53 B55        S  
879      1     11767  83.1583          C50        C  
887      0    112053  30.0000          B42        S  
889      0    111369  30.0000         C148        C  

[216 rows x 12 columns]

#### Average Age per Class

In [21]:
df_grp1=df.groupby(['Pclass'])

In [22]:
# Average Age per Pclass
df_grp1['Age'].mean()

Pclass
1    38.233441
2    29.877630
3    25.140620
Name: Age, dtype: float64

In [23]:
# Min/Max Age per Pclass
df_grp1['Age'].min()

Pclass
1    0.92
2    0.67
3    0.42
Name: Age, dtype: float64

In [24]:
df_grp1['Age'].max()

Pclass
1    80.0
2    70.0
3    74.0
Name: Age, dtype: float64

In [25]:
# Count
df_grp1['Age'].count()

Pclass
1    186
2    173
3    355
Name: Age, dtype: int64

#### Using agg() function

We can use the aggregate function to aggregrate results based on some column or our custom condition

In [27]:
df_grp2 = df.groupby(['Pclass']).agg({'Age':lambda x: np.mean(x)})
df_grp2

Age
Pclass           
1       38.233441
2       29.877630
3       25.140620

In [31]:
# Min/Max,count,Sum for each Pclass usign agg function
min_age = df.groupby(['Pclass']).agg({'Age':'min'}).rename(columns={'Age':'Min_Age'})
min_age

Min_Age
Pclass         
1          0.92
2          0.67
3          0.42

In [32]:
max_age = df.groupby(['Pclass']).agg({'Age':'max'}).rename(columns={'Age':'Max_Age'})
max_age

Max_Age
Pclass         
1          80.0
2          70.0
3          74.0

In [34]:
age_count = df.groupby(['Pclass']).agg({"Age":'count'}).rename(columns={'Age':'Age_Count'})
age_count

Age_Count
Pclass           
1             186
2             173
3             355

In [36]:
sum_age = df.groupby(['Pclass']).agg({'Age':'sum'}).rename(columns={'Age':'Sum_Age'})
sum_age

Sum_Age
Pclass         
1       7111.42
2       5168.83
3       8924.92

In [37]:
# Name of all passengers in that class
df_grp4 = df.groupby(['Pclass']).agg({'Name':lambda x: ','.join(x)})
df_grp4

Name
Pclass                                                   
1       Cumings, Mrs. John Bradley (Florence Briggs Th...
2       Nasser, Mrs. Nicholas (Adele Achem),Hewlett, M...
3       Braund, Mr. Owen Harris,Heikkinen, Miss. Laina...

In [44]:
# Creates a MultiOndex Dataframe
df_grp5 = df.groupby(['Pclass']).agg({'Age':['max','min']})
df_grp5

Age      
         max   min
Pclass            
1       80.0  0.92
2       70.0  0.67
3       74.0  0.42